docs/configuration.txt
Configuration
*************

The server side uses a `YAML <http://www.yaml.org/>`_ configuration file that defines the page layouts and allowed values. This file is usually called config.yaml.

Here is the general structure:

.. code-block:: yaml

  dpis:
    - 254
    - 190
    {...}

  ?maxSvgWidth: 2048 # set the maximum dimensions to 2048 points, this is useful when using MapServer and a maximum dimension is there
  ?maxSvgHeight: 2048
  ?integerSvg: false # the library in MapServer <= 5.6 does not support floating point values in the SVG coordinate space, set this to true if using a WMS that does not support floating point values in SVG coordinates

  ?ignoreCapabilities: false # assume client is correct and do not load capabilities.  This is not recommended to be used unless you it fails when false (false is default)
  ?disableLayersMerging: false # WMS layers with mergable parameters will be merged by default. Set this to true to disable attempts to merge.
  ?maxPrintTimeBeforeWarningInSeconds: 30 # if print jobs take longer than this then a warning in the logs will be written along with the spec.
  ?printTimeoutMinutes: 5 # The maximum time to allow a print job to take before cancelling the print job.  The default is 5 (minutes)
  ?formats:
    - pdf
    - png
    {...}

  scales:
    - 25000
    - 50000
    {...}

  hosts:
    - {HOST_WHITELIST_DEFINITION}
    {...}

  ?localHostForward: # For request on map.example.com we build an http request on localhost with the header Host=map.example.com, this is to don't pass throw the proxy.
  ?  from:
  ?    - map.example.com
  ?  https2http: True # For above hosts on request on https we build a request on http

  ?headers: ['Cookie', 'Referer'] # The header that will be copyed to the tiles http requests

  ?keys:
  ?  - !key
  ?    host: !dnsMatch
  ?      host: maps.google.com
  ?      port: 80
  ?    domain: !dnsMatch
  ?      host: localhost
  ?    key: 1234456
  ?    id: gmd-xyz


  ?fonts:
  ? - {PATH}

  ?globalParallelFetches: 5
  ?perHostParallelFetches: 5
  ?tilecacheMerging: false
  ?connectionTimeout: 30000           MF_V1.2
  ?socketTimeout: 180000              MF_V1.2
  ?outputFilename: Mapfish-print      MF_V1.2
  ?disableScaleLocking: false
  ?brokenUrlPlaceholder: default      MF_V2.0
  ?proxyBaseUrl: http://mapfishprint.org  MF_V2.0
  ?tmsDefaultOriginX: 0.0f   MF_V2.0
  ?tmsDefaultOriginY: 0.0f   MF_V2.0

  ?security:
  ?  - !basicAuth
  ?      matcher: !dnsMatch
  ?        host: www.camptocamp.com
  ?        post: 443
  ?      username: xyz
  ?      password: zyx
  ?      preemptive: true
  ?  - !basicAuth
  ?    username: abc
  ?    password: bca

  layouts:
     {LAYOUT_NAME}:
  ?   : Mapfish-print.pdf  MF_V1.2
  ?   metaData:
  ?     {METADATA_DEFINITION}
  ?   titlePage:
  ?     {PAGE_DEFINITION}
      mainPage:
  ?     rotation: false
        {PAGE_DEFINITION}
  ?   lastPage:
  ?     {PAGE_DEFINITION}
    {...}

Optional parts are shown with a question mark in the left margin. The question marks must not be put in the configuration file. Their default values is shown.

Note: Sets of values like DPI can be entered in one of two forms:
.. code-block:: yaml

  dpi: [1,2,3,...]

or

.. code-block:: yaml

  dpis:
    - 254
    - 190

A chosen DPI value from the above configuration is used in WMS GetMap requests as an added format_options (GeoServer) or map_resolution (MapServer) parameter. This is used for symbol/label-rescaling suitable for high resolution printouts, see `Geoserver format_options specification <http://docs.geoserver.org/latest/en/user/services/wms/vendor.html>`_ (Geoserver 2.1) and `MapServer defresolution keyword <http://mapserver.org/development/rfc/ms-rfc-55.html>`_ (MapServer 5.6) for more information.

In general, PDF dimensions and positions are specified in points. 72 points == 1 inch == 25.4 mm.

The list of {HOST_WHITELIST_DEFINITION} defines the allowed URLs for getting maps. Its format will be defined in `the next sub-section <configuration.html#host-whitelist-definition>`_.

The formats element lists the values formats that the server permits.  If omitted only 'pdf' is permitted.  If the single element '*' (quotes are required) is present then all formats that the server can produce can be requested.  The formats the server can produce depends to a large degree on how the Java is configured.  PDF is supported on all systems but for image output formats JAI and ImageIO is used which means both must be on the server for them to be available.  You can get the list of supported formats by running the standalone client with the --clientConfig flag enabled (you will need to supply a yaml config file as well).  If you are using the servlet then do a get info request to see the list of formats (with the '*' as the outputFormats parameter in the config file).

You can have as many layouts as you want. Their name must be unique and will be used on the client side. A layout can have a "titlePage" that will be added at the beginning of the generated document. It cannot contain any map. Same for the "lastPage", but for the end of the document. The "mainPage" section is mandatory and will be used once for each page requested. The details of a {PAGE_DEFINITION} section can be found `in another sub-section of this document <configuration.html#page-definition>`_.

If you want to let the user rotate the map (for a given layout), you have to set the "rotate" field to "true" in the corresponding "mainPage" section.

"globalParallelFetches" and "perHostParallelFetches" are used to tune the parallel loading of the map tiles/images. If you want to disable the parallel loading, set "globalParallelFetches" to 1.

New versions of tilecache added the support for merging multiple layers in a single WMS request. If you want to use this functionality, set the "tilecacheMerging" attribute to true.

"connectionTimeout" and "socketTimeout" (only since MapFish v1.2) can be used to tune the timeouts for reading tiles from map servers.

If the 'outputFilename' parameter is defined in the main body then that name will be used by the MapPrintServlet when sending the pdf to the client.  It will be the name of the file that the client downloads.  If the 'outputFilename' parameter is defined in a layout then that value will override the default name.  In both cases the .pdf is optional; if not present the server will append .pdf to the name.  In all cases the json request can override the filename defined in the configuration file by posting a 'outputFilename' attribute in the posted JSON. If the outputFilename has ${date}, ${time} or ${dateTime} in it, it will be replaced with the current date using the related DateFormat.get*Instance().format() method.  If a pattern is provided it will be passed to SimpleDataFormat for processing.  A few examples follow:

* outputFilename: "host-${yyyyMMdd}.pdf"   # results in host-20111213.pdf
* outputFilename: "host-${date}"           # results in host-Dec_13_2011.pdf (actual output depends on local of server)
* outputFilename: "host-${dateTime}"       # results in host-Dec_13_2011_1:10:50_PM.pdf (actual output depends on local of server)
* outputFilename: "host-${time}.pdf"       # results in host-1:11:14_PM.pdf (actual output depends on local of server)
* outputFilename: "host-${yyMMdd-hhmmss}"# results in host-111213-011154.pdf (actual output depends on local of server)

"disableScaleLocking" allows you to bypass the choosing of scale from the available factors, and simply use the suggested value produced inside MapBlock.java.

"brokenUrlPlaceholder" the placeholder image to use in the case of a broken url.  By default, when a url request fails, an error is thrown and the pdf process terminates.  However if this parameter is set then instead a placeholder image is returned.
Non-null values are:
  * "default" - use the system default image.
  * "throw" - throw an exception.
  * <url> - obtain the image from the supplied url.  If this url is broken then an exception will be thrown.  This can be anytype of valid url from a file url to https url.

"proxyBaseUrl" the optional url of the proxy between mapfish-print and the internet.  This is the url base that will be in the info.json response.  On occasion the url or port of the web server containing mapfish-print is not the server that is public to the internet and the requests are proxied to the mapfish-print webserver.  In this case it is important for the info.json request to return the public URL instead of the url of the webserver.

"tmsDefaultOriginX" By default this is null.  If non-null then TmsMapReader will use this as the origin x value if null then the origin will be derived from the maxExtent parameter.

"tmsDefaultOriginY" By default this is null.  If non-null then TmsMapReader will use this as the origin y value if null then the origin will be derived from the maxExtent parameter.


Security
--------
Both Keys and Security are options for accessing protected services.  Keys are currently for Google maps premium accounts and Security is for other types and is more general  Currently only BasicAuth is supported but other strategies can easily be added

.. code-block:: yaml

    security:
        - !basicAuth
            matcher: !dnsMatch
              host: www.camptocamp.com
              post: 443
            username: xyz
            password: zyx
            preemptive: true
        - !basicAuth
          username: abc
          password: cba

The above example has 2 security configuration.  Each option is tested (in order) to see if it can be used for the given URI and if it applies it is used to configure requests for the URI.  In the above example the first configuration will be used if the URI matches the hostmatcher provided if not then the second configuration will be applied.  The last configuration has no host matcher so it is applied to all URIs.

A basicAuth security configuration consists of 4 options

* matcher - a host matcher for determining which requests need the security to be applied
* username - username for basicauth
* password - password for basicauth
* preemptive - optional, but for cases where the credentials need to be sent without the challenge

Keys
----
Google maps currently requires a private key to be used (we only support users Google maps premium accounts).

The keys section allows a key to be mapped to hosts.  The hosts are identified with host matchers that are described in the <configuration.html#host-whitelist-definition> sub-section.

In addition a domain hostmatcher can be used to select a key based on the domain of the local server.  This can be useful if the same configuration is used in a test environment and a production environment with differing domains.  For example mapfish.org and mapfish.net.

Finally google maps (for example) requires a client id as well that is associated with the private key.  There for in the case of google premium services a legal key would be:

.. code-block:: yaml

  keys:
    - !key
      key: yxcvyxvcyxvyx
      id: gme-xxxcs

Thanks to the hosts and domain matcher it is possible to have a key for google maps and (for future proofing) a different key for a different service.

Fonts definition
-----------------

The "fonts" section is optional. It contains the path of the fonts you want to use. The entries can point to files (TTF, OTF, TTC, AFM, PFM) or directories. Don't point to directories containing too many files since it will slow down the start time. By default, PDF gives you access to the following fonts (Cp1252 encoding only):

* Courrier (-Bold, -Oblique, -BoldOblique)
* Helvetica (-Bold, -Oblique, -BoldOblique)
* Times (-Roman, -Bold, -Oblique, -BoldOblique)
* Symbol
* ZapfDingbats

Host whitelist definition
-------------------------

In this section, you can put as many entries as you want, even for the same type of filter. If at least one matches, the Map server can be used.

This section is not for defining which client can request maps. It is just here to avoid having the print module used as a proxy to access documents from computers behind firewalls.

There are 3 ways to whitelist a host.

Allowing every local services:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: yaml

    - !localMatch
      dummy: true

The "dummy" parameter is ignored, but mandatory to avoid a limitation in the YAML format.

Allowing by DNS name:
~~~~~~~~~~~~~~~~~~~~~

.. code-block:: yaml

    - !dnsMatch
      host: labs.metacarta.com

Allowing by IP address:
~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: yaml

      - !ipMatch
        ip: www.camptocamp.org
    ?   mask: 255.255.255.255

The "ip" parameter can be a DNS name that will be resolved or directly an IP address.

All the methods accept the following optional parameters:

* port: to limit to a certain TCP port
* pathRegexp: a regexp that must match the path part of the URL (before the '?').

Metadata definition
-------------------

Allow to add some metadata to the generated PDF. They are visible in acroread in the File->Properties menu.

The structure is like that:

.. code-block:: yaml

        metaData:
    ?     title: ''
    ?     author: ''
    ?     subject: ''
    ?     keywords: ''
    ?     creator: ''
    ?     supportLegacyReader: false

All fields are optional and can use global variables, as defined in the `Block definition <configuration.html#block-definition>`_ chapter. Page specific variables are not accessible.

Page definition
---------------

The structure is like that:

.. code-block:: yaml

          pageSize: A4
    ?     landscape: false
    ?     marginLeft: 40
    ?     marginRight: 40
    ?     marginTop: 20
    ?     marginBottom: 20
    ?     backgroundPdf: template.pdf
    ?     condition: null
    ?     header:
            height: 50
            items:
              - {BLOCK_DEFINITION}
              {...}
          items:
            - {BLOCK_DEFINITION}
            {...}
    ?     footer:
            height: 50
            items:
              - {BLOCK_DEFINITION}
              {...}

With the "condition" we can completely hide a page, same behavior than in block.

If "backgroundPdf" is specified, the first page of the given PDF file will be added as background of every page.

The "header" and "footer" sections are optional. If the "items" that are in the main section are too big, more pages are generated. The header and footer will be drawn on those pages as well.

Here is a short list of supported **pageSizes**:

+--------+-------+--------+
+ name   + width + height +
+========+=======+========+
+ LETTER + 612   + 792    +
+--------+-------+--------+
+ LEGAL  + 612   + 1008   +
+--------+-------+--------+
+ A4     + 595   + 842    +
+--------+-------+--------+
+ A3     + 842   + 1191   +
+--------+-------+--------+

The complete list can be found in http://api.itextpdf.com/itext/com/itextpdf/text/PageSize.html. If you want to use a custom page size, you can set **pageSize** to the width and the height separated by a space.

Block definition
----------------

The next sub-sections document the possible types of blocks.

In general, text values or URLs can contain values taken from the **spec** structure coming with the client's request. A syntax similar to shell is used: ${variableName}. If the current page is a **titlePage**, only the root values are taken. If it's a **mainPage**, the service will first look in the current **page** section then in the root values. Here is how to use this functionality::

    text: 'The value of mapTitle is: ${mapTitle}'

Some virtual variables can be used:

* ${pageNum}: The current page number.
* ${pageTot}: The total number of pages. Can be used only in text blocks.
* ${now}: The current date and time as defined by the machine's locale.
* ${now FORMAT}: The current date and time as defined by the FORMAT string. The syntax is here: http://java.sun.com/j2se/1.5.0/docs/api/java/text/SimpleDateFormat.html.
* ${configDir}: The absolute path to the directory of the configuration file.
* ${format PRINTF VAR}: Format the value of VAR using the provided `PRINTF format <http://java.sun.com/j2se/1.5.0/docs/api/java/util/Formatter.html#syntax>`_ (for example: %,d).


src/main/java/org/mapfish/print/config/Config.java
/*
 * Copyright (C) 2013  Camptocamp
 *
 * This file is part of MapFish Print
 *
 * MapFish Print is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * MapFish Print is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MapFish Print.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mapfish.print.config;

//import org.apache.commons.httpclient.HostConfiguration;
import com.codahale.metrics.MetricRegistry;
import org.apache.commons.httpclient.HttpClient;
import org.apache.commons.httpclient.HttpMethodBase;
import org.apache.commons.httpclient.MultiThreadedHttpConnectionManager;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.json.JSONException;
import org.json.JSONWriter;
import org.mapfish.print.Constants;
import org.mapfish.print.InvalidValueException;
import org.mapfish.print.PDFUtils;
import org.mapfish.print.ThreadResources;
import org.mapfish.print.config.layout.Layout;
import org.mapfish.print.config.layout.Layouts;
import org.mapfish.print.map.MapTileTask;
import org.mapfish.print.map.readers.MapReaderFactoryFinder;
import org.mapfish.print.map.readers.WMSServiceInfo;
import org.mapfish.print.output.OutputFactory;
import org.pvalsecc.concurrent.OrderedResultsExecutor;

import java.io.Closeable;
import java.net.InetSocketAddress;
import java.net.MalformedURLException;
import java.net.Proxy;
import java.net.ProxySelector;
import java.net.SocketException;
import java.net.URI;
import java.net.UnknownHostException;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.TreeSet;
//import org.mapfish.print.output.OutputFormat;

/**
 * Bean mapping the root of the configuration file.
 */
public class Config implements Closeable {
    public static final Logger LOGGER = Logger.getLogger(Config.class);

    private Layouts layouts;
    private TreeSet<Integer> dpis;
    private TreeSet<Double> scales;
    private String maxSvgWidth = "";
    private String maxSvgHeight = "";
    private double maxSvgW = Double.MAX_VALUE;
    private double maxSvgH = Double.MAX_VALUE;
    private Float tmsDefaultOriginX = null;
    private Float tmsDefaultOriginY = null;
    private boolean reloadConfig = false;

    private boolean ignoreCapabilities = false;
    private int maxPrintTimeBeforeWarningInSeconds = 30;
    private int printTimeoutMinutes = 5;
    private boolean disableLayersMerging = false;

    private ThreadResources threadResources;

    private boolean integerSvg = true;

    private List<String> overlayLayers = null;

    private TreeSet<String> fonts = null;
    private List<HostMatcher> hosts = new ArrayList<HostMatcher>();
    private HashMap localHostForward;
    private TreeSet<String> headers;
    private TreeSet<Key> keys;

    private int globalParallelFetches = 5;
    private int perHostParallelFetches = 5;
    private int socketTimeout = 40*60*1000; // 40 minutes //3*60*1000;
    private int connectionTimeout = 40*60*1000; // 40 minutes //30*1000;

    private boolean tilecacheMerging = false;
    private boolean disableScaleLocking = false;

    private List<SecurityStrategy> security = Collections.emptyList();

    private String outputFilename = "mapfish-print.pdf";

    /**
     * How much of the asked map we tolerate to be outside of the printed area.
     * Used only in case of bbox printing (use by the PrintAction JS component).
     */
    private static final double BEST_SCALE_TOLERANCE = 0.98;

    private TreeSet<String> formats; // private int svgMaxWidth = -1; private int svgMaxHeight = -1;

    private OutputFactory outputFactory;

    private MapReaderFactoryFinder mapReaderFactoryFinder;
    private String brokenUrlPlaceholder = Constants.ImagePlaceHolderConstants.THROW;
    private String proxyBaseUrl;
    private MetricRegistry metricRegistry;

    public Config() {
        hosts.add(new LocalHostMatcher());
    }

    public void setOutputFactory(OutputFactory outputFactory) {
        this.outputFactory = outputFactory;
    }

    public Layout getLayout(String name) {
        return layouts.get(name);
    }

    public void setLayouts(Layouts layouts) {
        this.layouts = layouts;
    }

    public void setDpis(TreeSet<Integer> dpis) {
        this.dpis = dpis;
    }

    public void setMaxSvgWidth(String maxSvgWidth) {
        this.maxSvgWidth = maxSvgWidth;
        this.maxSvgW = Double.parseDouble(maxSvgWidth);
    }
    public String getMaxSvgWidth() {
        return this.maxSvgWidth;
    }
    public void setMaxSvgHeight(String maxSvgHeight) {
        this.maxSvgHeight = maxSvgHeight;
        this.maxSvgH = Double.parseDouble(maxSvgHeight);
    }
    public String getMaxSvgHeight() {
        return this.maxSvgHeight;
    }

    public double getMaxSvgW() {
        return this.maxSvgW;
    }
    public double getMaxSvgH() {
        return this.maxSvgH;
    }

    public TreeSet<Integer> getDpis() {
        return dpis;
    }


    public void printClientConfig(JSONWriter json) throws JSONException {
        json.key("scales");
        json.array();
        for (Double scale : scales) {
            json.object();
            json.key("name").value("1:" + NumberFormat.getIntegerInstance().format(scale));
            json.key("value").value(scale.toString());
            json.endObject();
        }
        json.endArray();

        json.key("dpis");
        json.array();
        for (Integer dpi : dpis) {
            json.object();
            json.key("name").value(dpi.toString());
            json.key("value").value(dpi.toString());
            json.endObject();
        }
        json.endArray();

        json.key("outputFormats");
        json.array();
        if (outputFactory != null) {
            for (String format : outputFactory.getSupportedFormats(this)) {
                json.object();
                json.key("name").value(format);
                json.endObject();
            }
        }
        json.endArray();

        json.key("layouts");
        json.array();
        ArrayList<String> sortedLayouts = new ArrayList<String>();
        sortedLayouts.addAll(layouts.keySet());
        Collections.sort(sortedLayouts);
        for (int i = 0; i < sortedLayouts.size(); i++) {
            String key = sortedLayouts.get(i);
            json.object();
            json.key("name").value(key);
            layouts.get(key).printClientConfig(json);
            json.endObject();
        }
        json.endArray();
    }

    public void setScales(TreeSet<Number> scales) {
        // it is common for the config.yaml file to have integers only in the file
        this.scales = new TreeSet<Double>();
        for (Number scale : scales) {
            this.scales.add(scale.doubleValue());
        }
    }

    public boolean isScalePresent(double targetScale) {
        for (double scale : scales) {
            if (Math.abs(scale - targetScale) < 0.001) {
                return true;
            }
        }
        return false;
    }

    public void setHosts(List<HostMatcher> hosts) {
        this.hosts = hosts;
    }

    public void setFonts(TreeSet<String> fonts) {
        this.fonts = fonts;
    }

    public TreeSet<String> getFonts() {
        return fonts;
    }

    public void setKeys(TreeSet<Key> keys) {
        this.keys = keys;
    }

    public TreeSet<Key> getKeys() {
        TreeSet<Key> k = keys;
        if(k == null) k = new TreeSet<Key>();
        return k;
    }

    /**
     * Make sure an URI is authorized
     */
    public boolean validateUri(URI uri) throws UnknownHostException, SocketException, MalformedURLException {
        for (int i = 0; i < hosts.size(); i++) {
            HostMatcher matcher = hosts.get(i);
            if (matcher.validate(uri)) {
                if (LOGGER.isDebugEnabled()) {
                    LOGGER.debug("URI [" + uri + "] accepted by: " + matcher);
                }
                return true;
            }
        }
        return false;
    }

    /**
     * Called just after the config has been loaded to check it is valid.
     *
     * @throws InvalidValueException When there is a problem
     */
    public void validate() {
        if (this.threadResources == null) {
            throw new IllegalStateException("Config was not configured with a threadResources object.  Check spring configuration file " +
                                            "and make sure there is a ThreadResources bean");
        }
        if (layouts == null) throw new InvalidValueException("layouts", "null");
        layouts.validate();

        if (dpis == null) throw new InvalidValueException("dpis", "null");
        if (dpis.size() < 1) throw new InvalidValueException("dpis", "[]");

        if (scales == null) throw new InvalidValueException("scales", "null");
        if (scales.size() < 1) throw new InvalidValueException("scales", "[]");
        if (!(scales.iterator().next() instanceof Double)) throw new Error("scales should be converted to Doubles");

        if (hosts == null) throw new InvalidValueException("hosts", "null");
        if (hosts.size() < 1) throw new InvalidValueException("hosts", "[]");

        if (globalParallelFetches < 1) {
            throw new InvalidValueException("globalParallelFetches", globalParallelFetches);
        }
        if (perHostParallelFetches < 1) {
            throw new InvalidValueException("perHostParallelFetches", perHostParallelFetches);
        }

        if (socketTimeout < 0) {
            throw new InvalidValueException("socketTimeout", socketTimeout);
        }
        if (connectionTimeout < 0) {
            throw new InvalidValueException("connectionTimeout", connectionTimeout);
        }

        for (Key key : getKeys()) {
            key.validate();
        }

    }

    /**
     * @return The first scale that is bigger or equal than the target.
     */
    public double getBestScale(double target) {
        if (this.disableScaleLocking) {
            return target;
        }
        else {
            double finalTarget = target * BEST_SCALE_TOLERANCE;
            for (double scale : scales) {
                if (scale >= finalTarget) {
                    return scale;
                }
            }
            return scales.last();
        }

    }

    public OrderedResultsExecutor<MapTileTask> getMapRenderingExecutor() {
        return this.threadResources.getMapRenderingExecutor();
    }

    /**
     * Stop all the threads and stuff used for this config.
     */
    public synchronized void close() {
        WMSServiceInfo.clearCache();
    }


    public void setGlobalParallelFetches(int globalParallelFetches) {
        this.globalParallelFetches = globalParallelFetches;
    }

    public void setPerHostParallelFetches(int perHostParallelFetches) {
        this.perHostParallelFetches = perHostParallelFetches;
        System.getProperties().setProperty("http.maxConnections", Integer.toString(perHostParallelFetches));
    }

    /**
     * Get or create the http client to be used to fetch all the map data.
     */
    public HttpClient getHttpClient(URI uri) {
        MultiThreadedHttpConnectionManager connectionManager = getConnectionManager();
        HttpClient httpClient = new HttpClient(connectionManager);

        // httpclient is a bit pesky about loading everything in memory...
        // disabling the warnings.
        Logger.getLogger(HttpMethodBase.class).setLevel(Level.ERROR);

        // configure proxies for URI
        ProxySelector selector = ProxySelector.getDefault();

        List<Proxy> proxyList = selector.select(uri);
        Proxy proxy = proxyList.get(0);

        if (!proxy.equals(Proxy.NO_PROXY)) {
            InetSocketAddress socketAddress = (InetSocketAddress) proxy.address();
            String hostName = socketAddress.getHostName();
            int port = socketAddress.getPort();

            httpClient.getHostConfiguration().setProxy(hostName, port);
        }

        for(SecurityStrategy sec : security)
        if(sec.matches(uri)) {
            sec.configure(uri, httpClient);
            break;
        }
        return httpClient;
    }

    private MultiThreadedHttpConnectionManager getConnectionManager() {
        return this.threadResources.getConnectionManager();
    }

    public void setTilecacheMerging(boolean tilecacheMerging) {
        this.tilecacheMerging = tilecacheMerging;
    }

    public boolean isTilecacheMerging() {
        return tilecacheMerging;
    }

    public void setDisableScaleLocking(boolean disableScaleLocking) {
        this.disableScaleLocking = disableScaleLocking;
    }

    public boolean isDisableScaleLocking() {
        return disableScaleLocking;
    }

    public void setSocketTimeout(int socketTimeout) {
        this.socketTimeout = socketTimeout;
    }

    public void setConnectionTimeout(int connectionTimeout) {
        this.connectionTimeout = connectionTimeout;
    }
    
    public void setDisableLayersMerging(boolean disableLayersMerging) {
       this.disableLayersMerging = disableLayersMerging;
    }
    
    public boolean isDisableLayersMerging() {
        return disableLayersMerging;
    }

    public String getOutputFilename(String layoutName) {
        Layout layout = layouts.get(layoutName);
        String name = null;
        if(layout != null) {
            name = layout.getOutputFilename();
            if (name != null) {
                name = PDFUtils.getValueFromString(name); // get the string if it has ${now} or ${now DATEFORMAT} in it
            }
        }
        return name == null ? outputFilename : name;
    }

    public String getOutputFilename() {
        return outputFilename;
    }

    public void setOutputFilename(String outputFilename) {
        this.outputFilename = outputFilename;
    }

    public TreeSet<String> getFormats() {
        if(formats == null) return new TreeSet<String>();
        return formats;
    }

    public void setFormats(TreeSet<String> formats) {
        this.formats = formats;
    }

    public void setOverlayLayers(List<String> overlayLayers) {
      this.overlayLayers = overlayLayers;
    }

    public List<String> getOverlayLayers() {
      return overlayLayers;
    }

    /**
     * @return the integerSvg true if for example MapServer 5.6 or earlier is used where integers are put into the SVG
     */
    public boolean getIntegerSvg() {
        return integerSvg;
    }

    /**
     * @param integerSvg the integerSvg to set
     */
    public void setIntegerSvg(boolean integerSvg) {
        this.integerSvg = integerSvg;
    }

    /**
     * @return the reloadConfig
     */
    public boolean getReloadConfig() {
        return reloadConfig;
    }

    /**
     * @param reloadConfig the reloadConfig to set
     */
    public void setReloadConfig(boolean reloadConfig) {
        this.reloadConfig = reloadConfig;
    }

    public void setSecurity(List<SecurityStrategy> security) {
        this.security = security;
    }

    public void setMapReaderFactoryFinder(
            MapReaderFactoryFinder mapReaderFactoryFinder) {
        this.mapReaderFactoryFinder = mapReaderFactoryFinder;
    }

    public MapReaderFactoryFinder getMapReaderFactoryFinder() {
        return mapReaderFactoryFinder;
    }

    public void setLocalHostForward(HashMap localHostForward) {
        this.localHostForward = localHostForward;
        System.setProperty("sun.net.http.allowRestrictedHeaders", "true");
    }

    public boolean localHostForwardIsHttps2http() {
        if (localHostForward != null) {
            Object https2http = localHostForward.get("https2http");
            if (https2http != null && https2http instanceof Boolean) {
                return (Boolean)https2http;
            }
        }
        return false;
    }

    public boolean localHostForwardIsFrom(String host) {
        if (localHostForward != null) {
            Object from = localHostForward.get("from");
            if (from != null && from instanceof List) {
                return ((List)from).indexOf(host) >= 0;
            }
        }
        return false;
    }

    public void setHeaders(TreeSet<String> headers) {
        this.headers = headers;
    }

    public TreeSet<String> getHeaders() {
        return this.headers;
    }

    public void setBrokenUrlPlaceholder(String brokenUrlPlaceholder) {
        this.brokenUrlPlaceholder = brokenUrlPlaceholder;
    }

    public String getBrokenUrlPlaceholder() {
        return brokenUrlPlaceholder;
    }

    /**
     * Get the url of the proxy that resides between the servlet and the internet.
     * <p/>
     * This method may return null if there is no proxy between the servlet and the internet.
     *
     * @return the url of the proxy that resides between the servlet and the internet.
     */
    public String getProxyBaseUrl() {
        return proxyBaseUrl;
    }

    /**
     * Set the url of the proxy that resides between the servlet and the internet.
     * <p/>
     * This is an optional parameter as there may not be a proxy.
     *
     * @param proxyBaseUrl the url of the proxy as seen from the internet.
     */
    public void setProxyBaseUrl(String proxyBaseUrl) {
        this.proxyBaseUrl = proxyBaseUrl;
    }

    /**
     * Specify the default x - origin used by TMS map reader.  If null then the value will be derived from maxExtent.
     */
    public Float getTmsDefaultOriginX() {
        return tmsDefaultOriginX;
    }
    /**
     * Specify the default x - origin used by TMS map reader.  If null then the value will be derived from maxExtent.
     */
    public void setTmsDefaultOriginX(Float tmsDefaultOriginX) {
        this.tmsDefaultOriginX = tmsDefaultOriginX;
    }
    /**
     * Specify the default y - origin used by TMS map reader.  If null then the value will be derived from maxExtent.
     */
    public Float getTmsDefaultOriginY() {
        return tmsDefaultOriginY;
    }
    /**
     * Specify the default y - origin used by TMS map reader.  If null then the value will be derived from maxExtent.
     */
    public void setTmsDefaultOriginY(Float tmsDefaultOriginY) {
        this.tmsDefaultOriginY = tmsDefaultOriginY;
    }

    public void setThreadResources(ThreadResources threadResources) {
        this.threadResources = threadResources;
    }

    public boolean isIgnoreCapabilities() {
        return ignoreCapabilities;
    }

    public void setIgnoreCapabilities(boolean ignoreCapabilities) {
        this.ignoreCapabilities = ignoreCapabilities;
    }

    public int getMaxPrintTimeBeforeWarningInSeconds() {
        return maxPrintTimeBeforeWarningInSeconds;
    }

    public int getPrintTimeoutMinutes() {
        return printTimeoutMinutes;
    }

    public void setPrintTimeoutMinutes(int printTimeoutMinutes) {
        this.printTimeoutMinutes = printTimeoutMinutes;
    }

    public void setMaxPrintTimeBeforeWarningInSeconds(int maxPrintTimeBeforeWarningInSeconds) {
        this.maxPrintTimeBeforeWarningInSeconds = maxPrintTimeBeforeWarningInSeconds;
    }

    public void setMetricRegistry(MetricRegistry metricRegistry) {
        this.metricRegistry = metricRegistry;
    }

    public MetricRegistry getMetricRegistry() {
        return metricRegistry;
    }
}
